ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/VectorMapJoinAntiJoinLongOperator.java

```diff
@@ -0,0 +1,315 @@
+      for(VectorExpression ve : bigTableFilterExpressions) {
+        ve.evaluate(batch);
+      }
+
+      final int inputLogicalSize = batch.size;
+      if (inputLogicalSize == 0) {
+        return;
+      }
+
+      // Perform any key expressions.  Results will go into scratch columns.
+      if (bigTableKeyExpressions != null) {
+        for (VectorExpression ve : bigTableKeyExpressions) {
+          ve.evaluate(batch);
+        }
+      }
+
+      // The one join column for this specialized class.
+      LongColumnVector joinColVector = (LongColumnVector) batch.cols[singleJoinColumn];
+      long[] vector = joinColVector.vector;
+
+      // Check single column for repeating.
+      boolean allKeyInputColumnsRepeating = joinColVector.isRepeating;
+
+      if (allKeyInputColumnsRepeating) {
+        // All key input columns are repeating.  Generate key once.  Lookup once.
+        // Since the key is repeated, we must use entry 0 regardless of selectedInUse.
+        JoinUtil.JoinResult joinResult;
+        if (!joinColVector.noNulls && joinColVector.isNull[0]) {
+          // For anti join, if the right side is null then its a match.
+          joinResult = JoinUtil.JoinResult.MATCH;
+        } else {
+          long key = vector[0];
+          if (useMinMax && (key < min || key > max)) {
+            // Out of range for whole batch. Its a match for anti join. We can emit the row.
+            joinResult = JoinUtil.JoinResult.MATCH;
+          } else {
+            joinResult = hashSet.contains(key, hashSetResults[0]);
+            // reverse the join result for anti join.
+            if (joinResult == JoinUtil.JoinResult.NOMATCH) {
+              joinResult = JoinUtil.JoinResult.MATCH;
+            } else if (joinResult == JoinUtil.JoinResult.MATCH) {
+              joinResult = JoinUtil.JoinResult.NOMATCH;
+            }
+          }
+        }
+
+        // Common repeated join result processing.
+        if (LOG.isDebugEnabled()) {
+          LOG.debug(CLASS_NAME + " batch #" + batchCounter + " repeated joinResult " + joinResult.name());
+        }
+        finishAntiRepeated(batch, joinResult, hashSetResults[0]);
+      } else {
+        // NOT Repeating.
+
+        if (LOG.isDebugEnabled()) {
+          LOG.debug(CLASS_NAME + " batch #" + batchCounter + " non-repeated");
+        }
+
+        // We remember any matching rows in matches / matchSize.  At the end of the loop,
+        // selected / batch.size will represent both matching and non-matching rows for outer join.
+        // Only deferred rows will have been removed from selected.
+        int selected[] = batch.selected;
+        boolean selectedInUse = batch.selectedInUse;
+
+        int hashSetResultCount = 0;
+        int allMatchCount = 0;
+        int spillCount = 0;
+        long saveKey = 0;
+
+        // We optimize performance by only looking up the first key in a series of equal keys.
+        boolean haveSaveKey = false;
```

Comment:
Maybe rename to haveExistingKey? or HavePrevKey?

Reply:
I have kept the code same as Semi join. may be we can take up this refactoring as part of merging the code.